packages/@lwc/engine-core/src/framework/hooks.ts

```diff
@@ -192,12 +193,21 @@ export function createViewModelHook(elm: HTMLElement, vnode: VCustomElement) {
         setElementShadowToken(elm, shadowAttribute);
     }
     const def = getComponentInternalDef(ctor);
-    createVM(elm, def, {
+    const vm = createVM(elm, def, {
         mode,
         owner,
         tagName: sel,
         renderer: owner.renderer,
     });
+    assert.isFalse(
+        vm.renderMode === RenderMode.Shadow &&
+            vm.shadowMode === ShadowMode.Synthetic &&
+            owner.renderMode === RenderMode.Shadow &&
+            owner.shadowMode === ShadowMode.Native,
+        `<${vm.tagName}> (preferNativeShadow=false) cannot be composed inside of <${
+            getAssociatedVM(owner.elm).tagName
+        }> (preferNativeShadow=true)`
+    );
```

Comment:
Technically we don't need this `assert` since we already have the `assert` on `connectedCallback`, but I feel like it's good to throw early, before the element is even inserted into the DOM, if the template is obviously invalid.